### PR TITLE
Slow field calculator - Bugfix 20094 followup 096b4ce

### PR DESCRIPTION
**Pull request by elpaso**
@m-kuhn : much simpler approach

### CODE REVIEW

**Comment by m-kuhn** [comment]
Indeed much simpler :+1: 

One thing I couldn't check yet is cache invalidation, is that taken care of somehow?

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -7093,14 +7093,13 @@ void QgisApp::fieldCalculator()
 
 void QgisApp::attributeTable( QgsAttributeTableFilterModel::FilterMode filter )
 {
-  QgsVectorLayer *vectorLayer = qobject_cast<QgsVectorLayer *>( activeLayer() );
-  if ( !vectorLayer )
+  QgsVectorLayer *myLayer = qobject_cast<QgsVectorLayer *>( activeLayer() );
```
Comment (nirvn) [inline src/app/qgisapp.cpp]: I think we're hunting down the my prefix in variable name these days ;)

**Comment by elpaso** [comment]
This should do:
```
connect( mLayer, &QgsVectorLayer::dataChanged, this, &QgsVectorLayerCache::invalidate );
```

File: src/app/qgisapp.cpp (hunk 1 above)
Comment (elpaso) [reply src/app/qgisapp.cpp]: I know, I already fixed it but the I reverted the whole PR. Feel free to fix it.

**Comment by m-kuhn** [comment]
Is that part of the last pull request? I didn't get around to fully review that one in-depth.

**Comment by elpaso** [comment]
No, it was still there: https://github.com/qgis/QGIS/blob/master/src/core/qgsvectorlayercache.cpp#L41

**Comment by m-kuhn** [comment]
and the dataChanged signal is emitted when ...? (sorry I'm too lazy to look it up, all good if it works, just wanted to be sure!)

Something else, an edit command can also be done for a single change (e.g. geometry change), in this case a full reload is probably not warranted. Do we need some smartness here (either by having heuristics that decide when a reload is warranted based on e.g. amount of changed features or by an additional parameter to the edit command that the field calculator and co can set). Or is that taken care of in another way?

**Comment by elpaso** [comment]
Sorry: double checked and `invalidate()` is triggered by `updateFields`, not `dataChanged`, dataChanged is apparently only called when setting up dependencies so it's a no go.

So ... we probably still have the problem that not all edit commands will trigger an invalidate of the cache, if that was your concern.

I don't really like the idea of hardcoding something in the editingStarted signal.

Summarizing: the issue is that in case of massive updates we do not want the model reload to be triggered on every and each attributeValueChanged signals, but there is currently no way to know if an editCommand (or a rollback, that has exactly the same issue) is massive or not.

In case there is a change in the fields the current PR fixes the issue perfectly, because the cache is invalidated, but in case we have no field edits, I suppose the cache is not invalidated.

Thinking out loud now:

What about mimicking the Qt model behavior that passes a QModelIndex (a QgsFeatureIds in our case) to the dataChanged signal? 

This way we could invalidate the cache selectively (I didn't check if that's alredy implemented) by passing the feature ids that have been modified?

What I'm thinking of is to have a dataChanged signal emitted on every change of a feature, in addition to the individual atomic signals that are emitted for attr/geometry/add/delete changes.

This way, listeners could listen to the atomic signals or just to the final dataChanged signal that will only contain the feature ids of the changed features.

**Comment by m-kuhn** [comment]
> Summarizing: the issue is that in case of massive updates we do not want the model reload to be triggered on every and each attributeValueChanged signals, but there is currently no way to know if an editCommand (or a rollback, that has exactly the same issue) is massive or not.

I think that sums it up pretty nicely.

> What I'm thinking of is to have a dataChanged signal emitted on every change of a feature, in addition to the individual atomic signals that are emitted for attr/geometry/add/delete changes.

That's a good idea, but is that going to help us in the case of the attribute table which needs to be able to do both, incremental updates (single attribute changes via feature form should not require a request and just happen in memory) and massive updates where it's quicker to just hold back and reload in the end?

What I was thinking of is something similar to `QAbstractItemModel::beginInsertRows(parent, first, last)`, where code which modifies the data tells the model about the size of the change before actually doing it. Some metadata with `beginEditCommand` to allow the attribute table (or others) to decide if they want to go the incremental way or the invalidate cache way.

B.t.w I am not sure if the `dataChanged` signal is misleading in this case because (at least on dataprovider level) it says "When emitted from a QgsVectorDataProvider, any cached information such as feature ids should be invalidated." and does not contain any information for incremental updates (it's e.g. emitted when a shapefile repack happens and we have no idea how feature ids got reassigned).

**Comment by elpaso** [comment]
> > Summarizing: the issue is that in case of massive updates we do not want the model reload to be triggered on every and each attributeValueChanged signals, but there is currently no way to know if an editCommand (or a rollback, that has exactly the same issue) is massive or not.
> 
> I think that sums it up pretty nicely.
> 
> > What I'm thinking of is to have a dataChanged signal emitted on every change of a feature, in addition to the individual atomic signals that are emitted for attr/geometry/add/delete changes.
> 
> That's a good idea, but is that going to help us in the case of the attribute table which needs to be able to do both, incremental updates (single attribute changes via feature form should not require a request and just happen in memory) and massive updates where it's quicker to just hold back and reload in the end?
> 
> What I was thinking of is something similar to `QAbstractItemModel::beginInsertRows(parent, first, last)`, where code which modifies the data tells the model about the size of the change before actually doing it. Some metadata with `beginEditCommand` to allow the attribute table (or others) to decide if they want to go the incremental way or the invalidate cache way.
> 

Yeah, sorry, I wasn't clear at all, I meant a single signal that is always emitted at the end of any type of change, but only once, so for instance the field calculator would emit a lot of `attributeValueChanged` signals but only one `dataChanged`.
Btw, my idea is more or less the same as yours, you propose to tell the model the number of rows in advance, I propose to tell the list of changed feature ids at the end, because I don't think we can always know in advance how many rows have changed.

> B.t.w I am not sure if the `dataChanged` signal is misleading in this case because (at least on dataprovider level) it says "When emitted from a QgsVectorDataProvider, any cached information such as feature ids should be invalidated." and does not contain any information for incremental updates (it's e.g. emitted when a shapefile repack happens and we have no idea how feature ids got reassigned).

In any event, I'm non embarking myself in a signal refactoring now, I've added a model refresh (emitting model's `dataChanged( full model here)` at the end of the bulk change and this works well: the views are updated as expected (in a blink of an eye without hitting the provider) and the problem of the view now being updated in case of ediCommands not coming from the field calculator is also solved.

Please have a look, but I think we should merge as it is, it's probably not the perfect solution but still an improvement.

**Comment by m-kuhn** [comment]
> Btw, my idea is more or less the same as yours, you propose to tell the model the number of rows in advance, I propose to tell the list of changed feature ids at the end

I wonder how to do incremental updates with a signal only containing fids in the end. Doesn't the model need to know if it's an incremental or a batch update by the time the `attributeValueChanged` signal arrives?

> because I don't think we can always know in advance how many rows have changed.

but at least there is more chance to know the order of magnitude at the source of a change (e.g. field calculator) than on representation level (in the attribute table).

File: src/gui/attributetable/qgsattributetablemodel.cpp [hunk 2]
```diff
@@ -294,6 +301,9 @@ void QgsAttributeTableModel::fieldFormatterRemoved( QgsFieldFormatter *fieldForm
 
 void QgsAttributeTableModel::attributeValueChanged( QgsFeatureId fid, int idx, const QVariant &value )
 {
+  // Skip all updates if an edit command is running
+  if ( mBulkEditCommandRunning )
+    return;
   QgsDebugMsgLevel( QStringLiteral( "(%4) fid: %1, idx: %2, value: %3" ).arg( fid ).arg( idx ).arg( value.toString() ).arg( mFeatureRequest.filterType() ), 3 );
 
   for ( SortCache &cache : mSortCaches )
```
Comment (m-kuhn) [inline src/gui/attributetable/qgsattributetablemodel.cpp]: Is the sorting also updated properly after a bulk update?

Comment (m-kuhn) [reply src/gui/attributetable/qgsattributetablemodel.cpp]: and ... shouldn't that be qgis::as_const anyway?

Comment (elpaso) [reply src/gui/attributetable/qgsattributetablemodel.cpp]: No: it's not

Comment (elpaso) [reply src/gui/attributetable/qgsattributetablemodel.cpp]: What should I call to update the sorting?

**Comment by elpaso** [comment]
My reasoning is that we always defer updates at the end: if it's a small change then it does not make any difference if you defer updates or not, because it will be fast either ways. If it's a massive change, I think it will be useful to know if we've changed 60k rows or just 10, because in case they are 60k we just update the whole model, if they are (say) 10, we can choose to update the changed rows individually.

File: src/gui/attributetable/qgsattributetablemodel.cpp (hunk 2 above)
Comment (elpaso) [reply src/gui/attributetable/qgsattributetablemodel.cpp]: > and ... shouldn't that be qgis::as_const anyway?

I didn't look at that line, I trust the author ;)

```
2544391a9 src/gui/attributetable/qgsattributetablemodel.cpp (Matthias Kuhn      2018-05-04 15:34:34 +0200)     for ( SortCache &cache
```

**Comment by m-kuhn** [comment]
> if it's a small change then it does not make any difference if you defer updates or not, because it will be fast either ways

that's where I disagree, with slow network connections it may take a couple of 100 ms for a request which is not required because we already have everything in memory.

**Comment by elpaso** [comment]
Let's stick to this PR and the bug it solves without broadening the discussion too much.

**sorting**: this PR did nothing in regard to sorting, but does not introduce a regression either because sorting was not working anyway (check this in a commit before my recent commits to the attr table model and the calculator)

```
# Preparation:  open table, switch on editing and sort by ORDER column
# Feature id = 2 is the first row, ORDER value is 1, sorting is active

n [7]: layer.fields().lookupField('ORDER')
Out[7]: 4

In [8]: layer.changeAttributeValue(2, 4, 12345)
Out[8]: True

In [9]: layer.endEditCommand()
```
Table view is NOT updated, value is still 1, click in the cell and it becomes 12345, still no sorting!
See https://issues.qgis.org/issues/15976 for the changeset that broke it.

**cache invalidation** last commit takes care of the search caches and triggers a re-sort, but I hope does not regress to https://issues.qgis.org/issues/15976

**Comment by elpaso** [comment]
Let's give this a field trial. 
@luipir @gioman can you please check how this behaves in regards to https://issues.qgis.org/issues/15976 ? I tested it and I don't see a real problem here, especially if you use TAB to blur out of the field after editing.

**Comment by luipir** [comment]
building...

**Comment by m-kuhn** [comment]
Thanks @elpaso let's see how this works out!

**Comment by luipir** [comment]
confirmed it works tnx @elpaso ... but doing steps in https://issues.qgis.org/issues/15976 I found other 2 bugs in edit form @m-kuhn preparing issue!

**Comment by elpaso** [comment]
@luipir  thanks for checking!
@m-kuhn thanks for reviewing!

**Comment by gioman** [comment]
> Let's give this a field trial.
> @luipir @gioman can you please check how this behaves in regards to https://issues.qgis.org/issues/15976 ? I tested it and I don't see a real problem here, especially if you use TAB to blur out of the field after editing.

@elpaso seems ok to me!

**Pull request merged**